backend-rust/src/graphql_api.rs

```diff
@@ -1313,14 +1331,209 @@ struct Contract {
+            | Event::ContractResumed(_)
+            | Event::ContractUpgraded(_)
+            | Event::ContractUpdated(_) = event
+            {
+                Ok(())
+            } else {
+                Err(ApiError::InternalError(format!(
+                    "Not Transferred, ContractInterrupted, ContractResumed, ContractUpgraded, or \
+                     ContractUpdated event; Wrong event enum tag: {:?}",
+                    mem::discriminant(&event)
+                )))
+            }?;
+
+            let contract_event = ContractEvent {
+                contract_address_index: self.contract_address_index,
+                contract_address_sub_index: self.contract_address_sub_index,
+                sender: row.creator.into(),
+                event,
+                block_height: row.block_height,
+                transaction_hash: row.transaction_hash,
+                block_slot_time: row.block_slot_time,
+            };
+
+            contract_events.push(contract_event);
+            total_events_count += 1;
+        }
+
+        // Get the `init_transaction_event`.
+        if include_initial_event {
+            let row = sqlx::query!(
+            r#"
+SELECT
+    module_reference,
+    name as contract_name,
+    contracts.amount as amount,
+    blocks.slot_time as block_slot_time,
+    init_block_height as block_height,
+    transactions.events,
+    transactions.hash as transaction_hash,
+    accounts.address as creator,
+    version
+FROM contracts
+JOIN blocks ON init_block_height=blocks.height
+JOIN transactions ON init_block_height=transactions.block_height AND init_transaction_index=transactions.index
+JOIN accounts ON transactions.sender=accounts.index
+WHERE contracts.index=$1 AND contracts.sub_index=$2
+"#,
+                self.contract_address_index.0 as i64,
+                self.contract_address_sub_index.0 as i64
+            )
+            .fetch_optional(pool)
+            .await?
+            .ok_or(ApiError::NotFound)?;
+
+            let Some(events) = row.events else {
+                return Err(ApiError::InternalError("Missing events in database".to_string()));
+            };
+
+            let mut events: Vec<Event> = serde_json::from_value(events).map_err(|_| {
+                ApiError::InternalError("Failed to deserialize events from database".to_string())
+            })?;
+
+            if events.len() != 1 {
+                return Err(ApiError::InternalError(
+                    "Contract init transaction expects exactly one event".to_string(),
+                ));
+            }
+
+            // Contract init transactions have exactly one top-level event.
+            let event = events.swap_remove(0_usize);
+
+            match event {
+                Event::ContractInitialized(_) => Ok(()),
+                _ => Err(ApiError::InternalError(format!(
+                    "Not ContractInitialized event; Wrong event enum tag: {:?}",
+                    mem::discriminant(&event)
```

Comment:
The `Event` type does not implement `Debug` and since this error should never happen (also long we didn't corrupt the database), the enum_tag is mainly for internal debugging.

But long-term would be nice if somebody implements `Debug` on the `Event` type :).

Reply:
There are dependent types that don't implement it so will leave it as it is for now.

Reply:
I would just derive Debug on those too. It does no harm and should just be done.